java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -512,12 +518,15 @@ public void reset() {
 
   @Override
   public Field getField() {
-    List<org.apache.arrow.vector.types.pojo.Field> childFields = new ArrayList<>();
-    List<FieldVector> children = internalStruct.getChildren();
-    int[] typeIds = new int[children.size()];
-    for (ValueVector v : children) {
-      typeIds[childFields.size()] = v.getMinorType().ordinal();
-      childFields.add(v.getField());
+    int childCount = (int) Arrays.stream(typeFields).filter(field -> field != null).count();
+    List<org.apache.arrow.vector.types.pojo.Field> childFields = new ArrayList<>(childCount);
+    int[] typeIds = new int[childCount];
+    for (int i = 0; i < typeFields.length; i++) {
+      if (typeFields[i] != null) {
+        int curIdx = childFields.size();
+        typeIds[curIdx] = i;
+        childFields.add(typeFields[i]);
+      }
```

Comment:
Could you explain the motivation for this change? I'm not really following why it's necessary to add support for comparing value ranges?

Reply:
@TheNeuralBit Thanks a lot for your feedback.

Motivation for comparing value ranges:
For each type of vector, we need a way to compare individual elements/contiguous range of elements. This is implemented in class RangeEqualsVisitor, with each overload method processing one type of vectors.

Recently, we have introduced a new vector type: DenseUnionVector, so we need to add a new overload to support it. 

Motivation for this change:

The old implementation of this method mimics that of class UnionVector, as they are both union vectors, and share many similarities. When implementating this PR, we found this implementation caused some problems. So we fix it now.

The fundamental reason is that the type ids are defined differently for union and dense union vectors. For union vectors, the type id is the ordinal of the minor type of the child vector; whereas for dense union vectors, the type id is an incremental tiny int allocated by the vector when registering a child vector. 

When comparing vector elements, we need to compare their fields first, and the fields include child vector ids (type ids), so we need to change this method. 

If there are any other problems, please feel free to contact us. Thanks a lot.

Reply:
> The fundamental reason is that the type ids are defined differently for union and dense union vectors.

But why is there this difference? I don't actually see anything in the [specification](http://arrow.apache.org/docs/format/Columnar.html#union-layout) about a difference between type ids for sparse vs. dense unions. To me it looks like this PR is creating that difference.

It just seems like an unrelated change that's been pulled into this PR. An answer to this question would help me understand: Could you add support for comparing value ranges _without_ this change to type ids? If not, why not?

Reply:
@TheNeuralBit Good question.

You are right, as the difference is not mentioned in the specification. 

For Java, the dense union is aligning with the specification, while the sparse union is not. The dense union was implemented only recently, while the sparse union was implemented long ago. So it is not easy to fix the layout of sparse union, for the sake of backward compatibility. 

So it is the implementation of dense union that started introducing the difference, and this change makes the difference complete. I am afraid it is difficult to remove the difference, at least in the near future.

Reply:
Understood, thanks for the explanation